src/ess/dream/instrument_view.py

```
@@ -0,0 +1,135 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+from html import escape
+from typing import Union
+
+import plopp as pp
+import scipp as sc
+
+
+def _to_data_group(data: Union[sc.DataArray, sc.DataGroup, dict]) -> sc.DataGroup:
+    if isinstance(data, sc.DataArray):
+        data = sc.DataGroup({data.name or 'data': data})
+    elif isinstance(data, dict):
+        data = sc.DataGroup(data)
+    return data
+
+
+@pp.node
+def slice_range(da, trunc_range):
+    min_tr, max_tr = trunc_range
+    return da['tof', min_tr:max_tr].sum('tof')
+
+
+@pp.node
+def pre_process(da, dim):
+    dims = list(da.dims)
+    if dim is not None:
+        dims.remove(dim)
+    out = da.flatten(dims=dims, to='pixel')
+    sel = sc.isfinite(out.coords['x'])
```

What is `'x'`? The literal X-coord? Why do we need to select based on it?

---

This is because of the way we compute the coordinates for the pixels. We to a `mean` on the positions inside the bins, but some of the bins are empty (no events reached that pixel).
So some positions end up being NaN, and the 3D view complains about nans (nothing is rendered).

So I need some way to filter out NaNs. I agree that maybe the instrument view shouldn't silently drop NaNs, and maybe this should be done explicitly in the notebook?

Filtering on NaNs here was easy because I was internally flattening the data. Using advanced indexing only works using a 1D mask. But I think it makes sense to keep the data with logical dimensions in the notebook.
Which is why I wanted to filter here...

I don't know what the best solution is.

---

I thought about filling the empty pixels with a single event with zero weight, which would then appear as a dark pixel in the instrument view, but I then do not know what the positions of those pixels should be, because they are absent from the original table.

---

Makes sense.

From NeXus files we will (according to the current plan) get a `position` coord instead. Will you handle this as well?

---

I think that if I use the loader Jan-Lukas wrote, I get a position coord, so I will probably just filter on that.